botorch/acquisition/multi_step_lookahead.py

```diff
@@ -213,30 +249,46 @@ def _step(
             `i` is `f_i`. The last element of this list is considered the
             "inner sampler", which is used for evaluating the objective in case it is an
             MCAcquisitionObjective.
-        value_function_cls: The acquisition function class to be used as the value
-            function (providing the terminal value).
-        value_function_kwargs: A dictionary of arguments used in the value function
-            constructor.
+        stage_valfunc_cls: A list of acquisition function class to be used as the stage
+            value functions. Each element (except for the last one) can be None, in
+            which case a a zero stage value is assumed for the respective stage.
+        stage_valfunc_argfacs: A list of callables that map a `Model` and input tensor
+            `X` to a dictionary of kwargs for the respective stage value function
+            constructor. If None, only the standard `model`, `sampler` and
+            `objective` kwargs will be used.
+        inner_samplers: A list of `MCSampler` objects, each to be used in the stage
+            value function at the corresponding index.
         objective: The AcquisitionObjective under which the model output is evaluated.
+        running_val: As `batch_shape`-dim tensor containing the current running value.
         first_step: If True, this is considered to be the first step (resulting
             in not propagating gradients through the training inputs of the model).
 
     Returns:
         A `b`-dim tensor containing the multi-step value of the design `X`.
     """
-    if len(Xs) != len(samplers):
-        raise ValueError("Xs and samplers must have the same number elements")
+    if len(Xs) != len(samplers) + 1:
+        raise ValueError("Must have as many samplers as look-ahead steps")
 
-    # base case: no more fantasizing, compute value
-    if len(Xs) == 1:
-        kwargs: Dict[str, Any] = {"model": model, "objective": objective}
-        if issubclass(value_function_cls, MCAcquisitionFunction):
-            kwargs["sampler"] = samplers[0]
-        terminal_value_func = value_function_cls(**kwargs, **value_function_kwargs)
-        obj = terminal_value_func(X=Xs[0])
-        # shape of obj is (inner_mc_samples) x f_k x ... x f_1 x batch_shape
+    # compute stage value
+    StageValFuncClass = stage_valfunc_cls[0]
+    if StageValFuncClass is not None:
+        common_kwargs: Dict[str, Any] = {"model": model, "objective": objective}
+        if issubclass(StageValFuncClass, MCAcquisitionFunction):
+            common_kwargs["sampler"] = inner_samplers[0]
+        arg_fac = stage_valfunc_argfacs[0]
+        kwargs = arg_fac(model=model, X=Xs[0]) if arg_fac is not None else {}
```

Comment:
so in case of ExpectedImprovement, we need to pass in best_f. Seems currently it doesn't support that? But we could easily write a wrapper to assign best_f = train_input.max()

Reply:
So the idea is that `arg_fac` would provide the required kwargs, depending on the acquisition function type. For EI we would use `make_best_f` up at the top of the module. Then `best_f` would be passed into the constructor via `**kwargs`.

Reply:
ah, missed that part, that's nice!